subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.testing;
+
+import org.gradle.api.internal.tasks.testing.TestCompleteEvent;
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal;
+import org.gradle.api.internal.tasks.testing.TestStartEvent;
+import org.gradle.api.internal.tasks.testing.results.TestListenerInternal;
+import org.gradle.internal.logging.events.OperationIdentifier;
+import org.gradle.internal.operations.BuildOperationIdFactory;
+import org.gradle.internal.progress.BuildOperationCategory;
+import org.gradle.internal.progress.BuildOperationDescriptor;
+import org.gradle.internal.progress.BuildOperationListener;
+import org.gradle.internal.progress.BuildOperationState;
+import org.gradle.internal.progress.OperationFinishEvent;
+import org.gradle.internal.progress.OperationStartEvent;
+import org.gradle.internal.time.Clock;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Emitting build operations for tests
+ */
+public class TestListenerBuildOperationAdapter implements TestListenerInternal {
+    public static final TestOutputBuildOperationType.Details TEST_OUTPUT_DETAILS = new TestOutputBuildOperationType.Details() {
+    };
+    private final Map<TestDescriptor, TestBuildOperation> runningTests = new HashMap<TestDescriptor, TestBuildOperation>();
+    private final BuildOperationState testRootBuildOperation;
+    private final BuildOperationListener listener;
+    private final BuildOperationIdFactory buildOperationIdFactory;
+    private final Clock clock;
+
+    public TestListenerBuildOperationAdapter(BuildOperationState testRootBuildOperation, BuildOperationListener listener, BuildOperationIdFactory buildOperationIdFactory, Clock clock) {
+        this.testRootBuildOperation = testRootBuildOperation;
+        this.listener = listener;
+        this.buildOperationIdFactory = buildOperationIdFactory;
+        this.clock = clock;
+    }
+
+    @Override
+    public void started(final TestDescriptorInternal testDescriptor, TestStartEvent startEvent) {
+        if (testDescriptor.getParent() != null) {
```

Comment:
Why are we deviating from the model already used elsewhere where we wrap a “suite” around the whole tree?